relay-server/src/actors/relays.rs

```diff
@@ -79,49 +147,37 @@ impl RelayState {
     }
 }
 
-#[derive(Debug)]
-struct RelayInfoChannel {
-    sender: oneshot::Sender<Option<RelayInfo>>,
-    receiver: Shared<oneshot::Receiver<Option<RelayInfo>>>,
-}
+type FetchResult = Result<GetRelaysResult, HashMap<RelayId, Vec<Sender<GetRelayResult>>>>;
 
-impl RelayInfoChannel {
-    pub fn new() -> Self {
-        let (sender, receiver) = oneshot::channel();
-        RelayInfoChannel {
-            sender,
-            receiver: receiver.shared(),
-        }
-    }
-
-    pub fn send(self, value: Option<RelayInfo>) -> Result<(), Option<RelayInfo>> {
-        self.sender.send(value)
-    }
-
-    pub fn receiver(&self) -> Shared<oneshot::Receiver<Option<RelayInfo>>> {
-        self.receiver.clone()
-    }
-}
-
-pub struct RelayCache {
+#[derive(Debug)]
+pub struct RelayCacheService {
+    static_relays: HashMap<RelayId, RelayInfo>,
+    relays: HashMap<RelayId, RelayState>,
+    senders: HashMap<RelayId, Vec<Sender<GetRelayResult>>>,
+    fetch_channel: (mpsc::Sender<FetchResult>, mpsc::Receiver<FetchResult>),
     backoff: RetryBackoff,
+    delay: SleepHandle,
     config: Arc<Config>,
-    relays: HashMap<RelayId, RelayState>,
-    static_relays: HashMap<RelayId, RelayInfo>,
-    relay_channels: HashMap<RelayId, RelayInfoChannel>,
 }
 
-impl RelayCache {
+impl RelayCacheService {
     pub fn new(config: Arc<Config>) -> Self {
-        RelayCache {
-            backoff: RetryBackoff::new(config.http_max_retry_interval()),
+        Self {
             static_relays: config.static_relays().clone(),
-            config,
             relays: HashMap::new(),
-            relay_channels: HashMap::new(),
+            senders: HashMap::new(),
+            fetch_channel: mpsc::channel(1),
```

Comment:
We need to spawn a long(er) running background task to fetch relay infos and insert them into our map. It cannot hold mutable access to the map during that time, since there can be get requests in the meanwhile.

My solution is to have a dedicated mpsc to update the internal state when the fetch result is ready. Since there can only be a single fetch at any given time, this channel can have a bounded capacity of `1`.

An alternative solution to this would be an optional future similar to `SleepHandle`. I have an experimental implementation of that [here](https://github.com/getsentry/relay/compare/4881fd36a90a231cf513c4cc813afa802896be24...ref/experiment-spawn-handle), although I don't think that's necessarily better.